server/src/Icepeak/Server/WebsocketServer/MultiSubscription.hs

```diff
@@ -0,0 +1,440 @@
+
+instance Exception SubscriptionTimeout
+
+handleClient :: WS.Connection -> Core -> IO ()
+handleClient conn core = do
+  uuid <- Utils.newUUID
+  isDirty <- MVar.newMVar ()
+  subscriptions <- MVar.newMVar (HashMap.empty :: HashMap [Text] (MVar Value))
+
+  let
+    client =
+      Client
+        { clientConn = conn
+        , clientUuid = uuid
+        , clientCore = core
+        , clientIsDirty = isDirty
+        , clientSubscriptions = subscriptions
+        }
+
+    manageConnection =
+      Async.withAsync
+        (startUpdaterThread client)
+        -- This thread is only responsive to aync exceptions, and it ignores any other exception that may happen inside it and keeps going.
+        -- This makes sense for 2 reasons:
+        -- 1. `withAsync` will send a kill signal to the inner action ("updaterThread" in this case) when the outer action exists or throws an exception.
+        -- 2. Since we do not use the `Async` handle of the inner action, we do not call `wait` on it in the outer action.
+        --    `wait` is the mechanism that propagates any exception that might be thrown.
+        --     Since we do not use `wait`, if an exception is thrown inside the inner action, then it will just silently die.
+
+        -- It's important that the below action is the outer action of the `withAsync` as
+        -- we rely on the Exceptions the outer action throws for 4 reasons:
+        -- 1. For `withAsync` to kill the inner "updaterThread" thread when the outer action throws.
+        -- 2. To propagate the SubscriptionTimeout exception from the `withSubscribeTimeout`.
+        -- 3. To propagate the ConnectionException from the `receiveDataMessage` that is
+        --    thrown within the "messageHandlerThread".
+        -- 4. So that `onDisconnect` runs, see that `manageConnection` is in a `finally`.
+        ( const $
+            withSubscribeTimeout
+              client
+              (startMessageHandlerThread client)
+        )
+
+    -- Simply ignore connection errors, otherwise, Warp handles the exception
+    -- and sends a 500 response in the middle of a WebSocket connection, and
+    -- that violates the WebSocket protocol.
+    -- Note that subscribers are still properly removed by the finally below.
+
+    -- We can also ignore the websocket 'CloseRequest' exception because wai-websockets ensures
+    -- that the connection is closed when this IO action finishes, as this IO action is executing in a bracket.
+    -- https://github.com/yesodweb/wai/blob/c92d7b1993338fb0f91e0f5f34fb9678871028a0/wai-websockets/Network/Wai/Handler/WebSockets.hs#L98
+    handleConnectionError :: WS.ConnectionException -> IO ()
+    handleConnectionError _ = pure ()
+
+    -- Since we are in a bracket that closes the connection, let the action silently exit.
+    handleSubscriptionTimeout :: SubscriptionTimeout -> IO ()
+    handleSubscriptionTimeout _ = pure ()
+
+  -- Put the client in the subscription tree and keep the connection open.
```

Comment:
This comment seems slightly wrong: Unlike the single-subscription endpoint, here we don't put the client into the subscription tree, that should only happen with the first subscribe request we get.

(Though we do clean the remaining subscriptions when disconnecting).